### PR TITLE
Fix SSLHandshakeException while closing HTTPConduit

### LINKED ISSUE #24986: SSLHandshakeException occurs while closing HTTPConduit
The following SSLHandshakeException occurs while closing HTTPConduit:
javax.net.ssl.SSLHandshakeException:
PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
at sun.security.ssl.Alert.createSSLException(Alert.java:131)
at sun.security.ssl.TransportContext.fatal(TransportContext.java:324)
..
Caused by
at sun.security.ssl.SSLSocketImpl.readHandshakeRecord(SSLSocketImpl.java:1305)
at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:440)
at sun.net.www.protocol.https.HttpsClient.afterConnect(HttpsClient.java:559)
at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect(AbstractDelegateHttpsURLConnection.java:197)
at sun.net.www.protocol.http.HttpURLConnection.getOutputStream0(HttpURLConnection.java:1342)
at sun.net.www.protocol.http.HttpURLConnection.getOutputStream(HttpURLConnection.java:1317)
at sun.net.www.protocol.https.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:264)
at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.setupWrappedStream(URLConnectionHTTPConduit.java:318)
at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.handleHeadersTrustCaching(HTTPConduit.java:1394)
at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.onFirstWrite(HTTPConduit.java:1355)
at org.apache.cxf.transport.http.URLConnectionHTTPConduit$URLConnectionWrappedOutputStream.onFirstWrite(URLConnectionHTTPConduit.java:351)
at org.apache.cxf.io.AbstractWrappedOutputStream.write(AbstractWrappedOutputStream.java:47)
at org.apache.cxf.io.AbstractThresholdOutputStream.write(AbstractThresholdOutputStream.java:69)
at org.apache.cxf.transport.http.HTTPConduit$WrappedOutputStream.close(HTTPConduit.java:1409)

### PR DESCRIPTION
**Pull request by sawadood**
fixes #24986

### CODE REVIEW

**Comment by sawadood** [comment]
#spawn.fullfat.buckets=com.ibm.ws.jaxws.2.2.webcontainer_fat,com.ibm.ws.jaxws.2.2.webcontainer_fat_extended,com.ibm.ws.jaxws.cdi_fat,com.ibm.ws.jaxws.clientcontainer_fat,com.ibm.ws.jaxws.ejb_fat,com.ibm.ws.jaxws.X.wsat_fat,io.openliberty.xmlws.4.0.internal_fat,io.openliberty.ws.jaxws.global.handler.internal_fat,io.openliberty.jaxws.security_fat.ssl,io.openliberty.jaxws.security_fat.1,io.openliberty.jaxws.security_fat,com.ibm.ws.wssecurity_fat.wsscxf.1,com.ibm.ws.wssecurity_fat.wsscxf.2,com.ibm.ws.wssecurity_fat.wsscxf.3, com.ibm.ws.wssecurity_fat.wsscxf.4,com.ibm.ws.wssecurity_fat.wsscxf.saml.1,com.ibm.ws.wssecurity_fat.wsscxf.saml.2,com.ibm.ws.wssecurity_fat.wsscxf.saml.3,com.ibm.ws.wssecurity_fat.wsscxf.saml.4,com.ibm.ws.wssecurity_fat.wsscxf.saml.5,com.ibm.ws.wssecurity_fat.wsscxf.saml.6

**Comment by sawadood** [comment]
#build 
#spawn.fullfat.buckets=com.ibm.ws.jaxws.2.2.webcontainer_fat,com.ibm.ws.jaxws.2.2.webcontainer_fat_extended,com.ibm.ws.jaxws.cdi_fat,com.ibm.ws.jaxws.clientcontainer_fat,com.ibm.ws.jaxws.ejb_fat,com.ibm.ws.jaxws.X.wsat_fat,io.openliberty.xmlws.4.0.internal_fat,io.openliberty.ws.jaxws.global.handler.internal_fat,io.openliberty.jaxws.security_fat.ssl,io.openliberty.jaxws.security_fat.1,io.openliberty.jaxws.security_fat,com.ibm.ws.wssecurity_fat.wsscxf.1,com.ibm.ws.wssecurity_fat.wsscxf.2,com.ibm.ws.wssecurity_fat.wsscxf.3, com.ibm.ws.wssecurity_fat.wsscxf.4,com.ibm.ws.wssecurity_fat.wsscxf.saml.1,com.ibm.ws.wssecurity_fat.wsscxf.saml.2,com.ibm.ws.wssecurity_fat.wsscxf.saml.3,com.ibm.ws.wssecurity_fat.wsscxf.saml.4,com.ibm.ws.wssecurity_fat.wsscxf.saml.5,com.ibm.ws.wssecurity_fat.wsscxf.saml.6,com.ibm.ws.wsat_fat.2,com.ibm.ws.wsat_fat.3,com.ibm.ws.wsat_fat.4,com.ibm.ws.wsat_fat.5,com.ibm.ws.wsat_fat.6,com.ibm.ws.wsat_fat.7,com.ibm.ws.wsat_fat.db,com.ibm.ws.wsat.common_fat,com.ibm.ws.wsat.concurrent_fat,com.ibm.ws.wsat.migration_fat.1,com.ibm.ws.wsat.migration_fat.2,com.ibm.ws.wsat.migration_fat.3,com.ibm.ws.wsat.migration_fat.4,com.ibm.ws.wsat.migration_fat.5,com.ibm.ws.wsat.recovery_fat.lps,com.ibm.ws.wsat.recovery_fat.multi.1,com.ibm.ws.wsat.recovery_fat.multii.2,com.ibm.ws.wsat.recovery_fat.multii.3,com.ibm.ws.wsat.recovery_fat.multii.4,com.ibm.ws.wsat.recovery_fat.single.1,com.ibm.ws.wsat.recovery_fat.single.2,com.ibm.ws.jaxws.2.2.webcontainer_fat,com.ibm.ws.jaxws.2.2.webcontainer_fat_extended,com.ibm.ws.jaxws.cdi_fat,com.ibm.ws.jaxws.clientcontainer_fat,com.ibm.ws.jaxws.ejb_fat,com.ibm.ws.jaxws.X.wsat_fat,com.ibm.ws.javaee.dd.model_fat,com.ibm.ws.microprofile.rest.client.FT_fat,com.ibm.ws.microprofile.rest.client11_fat_tck,com.ibm.ws.microprofile.rest.client12_fat_tck,com.ibm.ws.microprofile.rest.client13_fat_tck,com.ibm.ws.microprofile.rest.client14_fat_tck,com.ibm.ws.microprofile.rest.client_fat,com.ibm.ws.microprofile.rest.client_fat_tck

**Comment by sawadood** [comment]
Personal build has 11 failures, all these failures have the same error ""invalid test bucket"" and not related to my fix.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (BerksanAtes) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:162]: Great idea to add SSL context check. It might be good to add trace showing when context is null?

**Review by BerksanAtes — APPROVED** [review]
It looks great! Especially, I believe  added traces will helps us tremendously!!

**Comment by sawadood** [comment]
#run-libby-bot

**Comment by sawadood** [comment]
Tested this fix locally also and did not find any issues.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
Comment (WhiteCat22) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java:132]: Can you confirm that `org.apache.cxf.rest.message` is present on MP Rest Client requests in addition to vanilla JAX-RS request please?

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (WhiteCat22) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:177]: `// Liberty Change End`

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (WhiteCat22) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:160]: Move this `// Liberty Change End` to after line 178

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/SSLUtils.java
Comment (WhiteCat22) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/SSLUtils.java:77]: Remove TS012061109

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/SSLUtils.java
Comment (WhiteCat22) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/SSLUtils.java:56]: remove TS012061109

**Comment by WhiteCat22** [comment]
@sawadood I don't know why it's not allowing my to block the merge with "Request Changes", but I would like to see my comments addressed before you merge, thanks!

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java:412]: It seems to me that this code section might be incorrect (but would not cause a problem in most cases).  They are attempting to get the "Error Stream".  Shouldn't the if check here be `if (in != null)`?

Also, I'd suggest changing your log message to `"Ignoring unexpected exception in getInputStream() when an error stream is present: "`

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:99]: Minor:  Change `get` to `create`

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:191]: Suggestion:  Remove the if check and change the line to the following so you'll know if it is null or not more easily:
`            LOG.fine(""Hostname verifier obtained from SSLUtils.getHostnameVerifier: " + (verifier==null? "null": verifier.getClass().getCanonicalName()));
`

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:195]: same suggestion here as above.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (jim-krueger) [inline dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:204]: same comment

**Review by jim-krueger — COMMENTED** [review]
Looks good.  Just a couple comments/suggestions.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java
Comment (WhiteCat22) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/https/HttpsURLConnectionFactory.java:191]: I think we should leave the if check, but only print a message if the verifier is null.

File: dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java
Comment (sawadood) [reply dev/com.ibm.ws.org.apache.cxf.cxf.rt.transports.http.3.2/src/org/apache/cxf/transport/http/URLConnectionHTTPConduit.java:132]: Yes, I have tested that org.apache.cxf.rest.message is present on MP REST client requests.

**Pull request merged**
